Products upload

K-Hir-pk18 opened the pull request:
WHAT
商品出品機能

WHY
登録者が商品を出せるようにするため

K-Hir-pk18 committed:
first commit

K-Hir-pk18 committed:
商品出品機能

WHAT
商品出品機能

WHY
登録者が商品を出せるようにするため

yuri240 commented:
レビュー依頼ありがとうございます！

こちらのプルリクエストですが、
実際の画面や機能ができているか確認させていただきたく存じます。

つきましては、下記キャプチャ動画・画像をご提出いただけますでしょうか？
```
・新規登録に成功する動画
・新規登録に失敗し、エラーが表示される動画
・ログインに成功する動画
・ログインに失敗し、エラーが表示される動画
・ログアウトに成功する動画
・単体テストが通っていることが確認できる画像
```
お手数をおかけいたしますが、よろしくお願いいたします。

K-Hir-pk18 commented:
yuri240様
ご担当者様

K-Hir-pk18です。
ご対応ありがとうございます。

以下にGYAZOのキャプチャを添付します。
ご確認よろしくお願い致します。


・新規登録に成功する動画
https://gyazo.com/4e9b761e0c11545339d6cca381b6305c

・新規登録に失敗し、エラーが表示される動画
https://gyazo.com/42dbd8e9313a24bdc44ca5e3cd437f06

・ログインに成功する動画
https://gyazo.com/299b63a812db221fcfed7149348bed29

・ログインに失敗し、エラーが表示される動画
https://gyazo.com/cb2b1e8702761eeca3428c3928bd2c41

・ログアウトに成功する動画
https://gyazo.com/d6083586db2c7cf29ed390bbdfae86ee

・単体テストが通っていることが確認できる画像
https://gyazo.com/3697797fe4802193ded6a16e3f087741

Hide-Morishita commented on app/controllers/application_controller.rb:
```diff
@@ -1,7 +1,7 @@
   #before_action :basic_auth
   before_action :configure_permitted_parameters, if: :devise_controller?
-
+  before_action :authenticate_user!, except: [:index, :show]
```
```ruby
before_action :authenticate_user!, except: [:index, :show]
```
こちらの記述ですが、applicaciton_controllerに記述するのではなく、products_controllerに記述しましょう。
applicaciton_controllerに記述してしまうと、継承しているすべてのcontrollerに適用されてしまうためです。

Hide-Morishita commented on app/controllers/products_controller.rb:
```diff
@@ -22,25 +29,28 @@ def index
-      #redirect_to action: :index
-    #end
-  #end
+  def move_to_index
+    unless user_signed_in?
+      redirect_to action: :index
+    end
+  end
```
こちらは、`:authenticate_user!`を使用することでログインしていないユーザーは、ログインページに遷移させることが可能ですので、不要な記述になります。
削除しておきましょう。

Hide-Morishita commented on app/javascript/profit.js:
```diff
@@ -0,0 +1,9 @@
+window.addEventListener('load', () => {
+  const priceInput = document.getElementById("item-price");
+  priceInput.addEventListener("input", () => {
+    const inputValue = priceInput.value;
+    const addTaxDom = document.getElementById("profit");
+    addTaxDom.innerHTML = inputValue - Math.floor(inputValue * 0.1)
+  console.log(inputValue);
+  });
+  })
```
販売手数料の計算ですが、1つのJsファイルで記述することが可能です。
1つのファイルで記述してみましょう。

Hide-Morishita commented on app/views/products/new.html.erb:
```diff
@@ -5,7 +5,7 @@
-    <%= form_with local: true do |f| %>
+    <%=form_with model:@product, local: true do |f| %>
 
     <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
```ruby
<%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
こちらは、実装案内用のコメントアウトですので削除しておきましょう。

Hide-Morishita commented on spec/models/product_spec.rb:
```diff
@@ -1,5 +1,94 @@
+        @product.send_span_id = nil
+        @product.valid?
+        expect(@product.errors.full_messages).to include("Send span can't be blank")
+        end
```
active_hashカラム(商品の状態, 配送料の負担, 発送元の地域, 発送までの日数, カテゴリー)には、「0を選択している場合登録できない」ようにバリデーションを記述しています。
「0を選択している場合登録できない」ことを確認するテストコードも記述しましょう。

Hide-Morishita commented on spec/models/product_spec.rb:
```diff
@@ -1,5 +1,94 @@
+        expect(@product.errors.full_messages).to include("Price can't be blank")
+        end
+
+      it '価格の範囲が、¥300以下だと出品できない' do
+        @product.price = '250'
+        @product.valid?
+        expect(@product.errors.full_messages).to include("Price is out of setting range")
+      end
+      
+      it '価格の範囲が、¥9,999,999以上だと出品できない' do
+        @product.price = '99999999'
+        @product.valid?
+        expect(@product.errors.full_messages).to include("Price is out of setting range")
+      end
```
```ruby
@product.price = '250'
@product.price = '99999999'
```
priceはinteger型ですので、値は全て文字列ではなく数値にしましょう。
（シングルクォーテーションを外しましょう）

販売価格に関してですが、より正確なテストを行うために、下記テストコードも実装しましょう。
```
・半角英数混合では出品できないこと
・半角英語だけでは出品できないこと
```

Hide-Morishita reviewed (commented):
お待たせしました、レビュー依頼ありがとうございます。
コメントいたしましたので、ご確認をお願いいたします。
前回ユーザー管理機能のキャプチャ動画をお願いしておりましたが、
今回は出品機能かと存じますので下記のキャプチャ動画をご提出していただけますでしょうか。
お手数をおかけいたしますがよろしくお願いいたします。
```
・出品に成功する動画
・出品に失敗しエラーが表示される動画
・ログイン状態のユーザーだけが、商品出品ページへ遷移できる動画
・ログアウト状態のユーザーは、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
・金額が入力されたと同時に手数料と利益が表示される動画
・テスト結果の画像
```
「Gyazo　使い方」とgoogleで検索し、ツールによって生成されたURLをこのPRに貼っていただければ幸いです。
ご不明な点がございましたら、コネクトにてご質問いただければと存じます。

K-Hir-pk18 commented:
Hide-Morishita様

チェックいただきありがとうございます。
内容確認しました。

・出品に成功する動画
https://gyazo.com/8fe5dd7ee07225659482138d6a23ff59

・出品に失敗しエラーが表示される動画
https://gyazo.com/aa828d1ac0493c16a3335f647c53a282

・ログイン状態のユーザーだけが、商品出品ページへ遷移できる動画
https://gyazo.com/3d247b45ef7adae783411ed97e7db00a

・ログアウト状態のユーザーは、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
https://gyazo.com/c9ea965a4b96636c31c9b1fa2cedc50b

・金額が入力されたと同時に手数料と利益が表示される動画
https://gyazo.com/14d7c95fb43185c26730583643e373b7

・テスト結果の画像
https://gyazo.com/914f5e8be7f6d3c7d165f633aa4de70b

K-Hir-pk18 committed:
レビュー対応

WHAT

・ before_action :authenticate_user!
　　→application.rb　から　product.rb　へ移動

・move_to_index の削除
　　→private method と before action

・Jsファイルの統合

・products/newのviewファイル
　　→コメントアウトの削除

・product_spec.rb
　　→カテゴリ選択のパラメータ０テスト追加
　　→priceのシングルクォーテーション削除

WHY
レビュー対応

K-Hir-pk18 commented:
ご担当者様


よろしくお願いいたします。

kuu0123 commented on app/controllers/products_controller.rb:
```diff
@@ -1,17 +1,24 @@
-  #end
+  def create
+    @product = Product.new(product_params)
+    if @product.valid?
+      @product.save
+      redirect_to root_path
+    else
+      render :new
+    end
```
```ruby
    if @product.valid?
      @product.save
```
`save`メソッドにはバリデーションをチェックする機能があるため、こちらは`@product.save`で条件分岐をするようにしておきましょう。

K-Hir-pk18 replied:
kuu0123様

レビューありがとうございます。
的確なチェックで、こっちも く〜💦 って感じです。
あとアイコンかわいいです。

…再チャレンジします(｀・ω・´)

K-Hir-pk18 replied:
ご担当者様

よろしくお願いいたします。

kuu0123 reviewed (commented):
レビュー依頼ありがとうございます！

修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします！

また、差分が出ていなかったためこちらに1点コメントさせていただきます。
products/new.html.erbですが、エラーメッセージの表示の箇所が削除されてしまっているようです。
キャプチャ動画を拝見してもエラーメッセージの確認ができませんでしたので、お手数ですが一度ご確認いただけますでしょうか。

よろしくお願いいたします。

K-Hir-pk18 committed:
レビュー対応２

WHAT
・products/new.html.erb
　　→エラーメッセージ

・products_controller.rb
　　→createアクションの@produtct.saveの記述位置変更

WHY
・レビューコメント対応

・より良いアプリとコードにするため！

商品出品エラー２
https://gyazo.com/258fe3316cfd1cf0b655f10b3a488b12

K-Hir-pk18 commented:
ご担当者様

よろしくお願いいたします。

kuu0123 commented on app/controllers/products_controller.rb:
```diff
@@ -1,17 +1,25 @@
-  #def create
-    #Product.create(product_params)
-  #end
+  def create
+    @product = Product.new(product_params)
+    @product.save
+        if @product.valid?
+      redirect_to root_path
+    else
+      render :new
+    end
+  end
```
```ruby
@product.save
        if @product.valid?
      redirect_to root_path
    else
      render :new
    end
```
こちらの条件分岐ですが、まだ改善できそうですね！
```ruby
    @product.save
        if @product.valid?
```
という記述だと、`@product.saveを行ってからバリデーションをチェックする`という形になります。
今回は、`saveができたかどうか`で条件分岐を行い、trueならトップページへ、falseなら再度出品のビューに飛ばしたいですね。
saveを使用する場合は、valid?は不要な記述になりますので、再度挑戦してみましょう！

K-Hir-pk18 replied:
kuu0123様

チェック頂きありがとうございます。
またしても　くぅ〜💦　な感じです。

次は　く〜✨✨　を目指します。

kuu0123 reviewed (commented):
レビュー依頼ありがとうございます！

修正点確認させていただきました。
追加で1点コメントいたしましたので、ご確認お願いします！

K-Hir-pk18 committed:
レビュー対応３

WHAT
・product_controllerの改良

```
  def create
    @product = Product.new(product_params)
    if @product.save
      redirect_to root_path
    else
      render :new
    end
  end
```

WHY
・レビュー対応

・良いコードにするため

・優秀なエンジニアに近づくため！

K-Hir-pk18 commented:
ご担当者様

よろしくお願いいたします。

K-Hir-pk18 committed:
rubocop作動

WHAT
・rubocop
　　対象に作動させた

WHY
・レギュレーション上
（可読性の向上など）

K-Hir-pk18 commented:
ご担当者様


よろしくお願いいたします。

kuu0123 commented:
レビュー依頼ありがとうございます！

商品出品機能、LGTMとさせていただきます！実装お疲れさまでした！

こちらのブランチは忘れずにMergeいただき、ローカルのMasterにPullした後に次の実装のためのブランチを切るようにしましょう。
引き続き実装頑張ってください！

K-Hir-pk18 commented:
kuu0123様

３回もレビューありがとうございます。
ついに、　く〜✨✨　と言えました。
ありがとうございます！
